From 7b0020382af964e97da0ec495011c21987fda729 Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:36:57 -0400
Subject: [virt] kvm/iommu: hva align mapping page size

Message-id: <1406040016-3289-23-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86101
O-Subject: [RHEL7 PATCH v2 022/221] KVM: IOMMU: hva align mapping page size
Bugzilla: 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Greg Edwards <gedwards@ddn.com>

When determining the page size we could use to map with the IOMMU, the
page size should also be aligned with the hva, not just the gfn.  The
gfn may not reflect the real alignment within the hugetlbfs file.

Most of the time, this works fine.  However, if the hugetlbfs file is
backed by non-contiguous huge pages, a multi-huge page memslot starts at
an unaligned offset within the hugetlbfs file, and the gfn is aligned
with respect to the huge page size, kvm_host_page_size() will return the
huge page size and we will use that to map with the IOMMU.

When we later unpin that same memslot, the IOMMU returns the unmap size
as the huge page size, and we happily unpin that many pfns in
monotonically increasing order, not realizing we are spanning
non-contiguous huge pages and partially unpin the wrong huge page.

Ensure the IOMMU mapping page size is aligned with the hva corresponding
to the gfn, which does reflect the alignment within the hugetlbfs file.

Reviewed-by: Marcelo Tosatti <mtosatti@redhat.com>
Signed-off-by: Greg Edwards <gedwards@ddn.com>
Cc: stable@vger.kernel.org
Signed-off-by: Gleb Natapov <gleb@redhat.com>
(cherry picked from commit 27ef63c7e97d1e5dddd85051c03f8d44cc887f34)
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/virt/kvm/iommu.c b/virt/kvm/iommu.c
index c7d9ce1..0df7d4b 100644
--- a/virt/kvm/iommu.c
+++ b/virt/kvm/iommu.c
@@ -103,6 +103,10 @@ int kvm_iommu_map_pages(struct kvm *kvm, struct kvm_memory_slot *slot)
   while ((gfn << PAGE_SHIFT) & (page_size - 1))
    page_size >>= 1;
 
+  /* Make sure hva is aligned to the page size we want to map */
+  while (__gfn_to_hva_memslot(slot, gfn) & (page_size - 1))
+   page_size >>= 1;
+
   /*
    * Pin all pages we are about to map in memory. This is
    * important because we unmap and unpin in 4kb steps later.
-- 
1.7.1